src/oni2-keyboard-layout/stubs/keyboard-layout-mac.c

```diff
@@ -0,0 +1,243 @@
+    // If we didn't get anything back from the OCaml runtime, return early.
+    if (callbackListRef == NULL) {
+      NSLog(@"Unable to acquire callback list reference!");
+      return;
+    }
+  }
+
+  NSString *nsLayout = oni2_priv_GetCurrentKeyboardLayout();
+  NSString *nsLanguage = oni2_priv_GetCurrentKeyboardLanguage();
+
+  value vLayout = caml_copy_string([nsLayout UTF8String]);
+  value vLanguage = caml_copy_string([nsLanguage UTF8String]);
+  value args[] = {vLanguage, vLayout};
+
+  value callbackList = UNWRAP_REF(*callbackListRef);
+
+  caml_c_thread_register();
+  caml_acquire_runtime_system();
+
+  // Loop through the callback list and call all of them with the langusage and layout
+  while (callbackList != Val_emptylist) {
+    value callback = Field(callbackList, 0);
+    caml_callbackN(callback, 2, args);
+    callbackList = Field(callbackList, 1);
+  }
+
+  caml_release_runtime_system();
+}
+
+CAMLprim value oni2_KeyboardLayoutInit() {
+  CAMLparam0();
+
+  CFNotificationCenterAddObserver(
+    CFNotificationCenterGetDistributedCenter(),
+    NULL,
+    notificationHandler,
+    kTISNotifySelectedKeyboardInputSourceChanged,
+    NULL,
+    CFNotificationSuspensionBehaviorDeliverImmediately
+  );
+
+  CAMLreturn(Val_unit);
+}
+
+CAMLprim value oni2_KeyboardLayoutGetCurrentLayout() {
+  CAMLparam0();
+  CAMLlocal1(vLayout);
+
+  vLayout = caml_copy_string([oni2_priv_GetCurrentKeyboardLayout() UTF8String]);
+
+  CAMLreturn(vLayout);
+}
+
+CAMLprim value oni2_KeyboardLayoutGetCurrentLanguage() {
+  CAMLparam0();
+  CAMLlocal1(vLanguage);
+
+  vLanguage = caml_copy_string([oni2_priv_GetCurrentKeyboardLanguage() UTF8String]);
+
+  CAMLreturn(vLanguage);
+}
+
+/* Given a keyboard layout, a key code, and modifiers,
+   fill a buffer of UTF-16 characters with the resulting
+   string
+ */
+size_t characterForNativeCode(
+  const UCKeyboardLayout *keyboardLayout,
+  uint16_t virtualKeyCode,
+  EventModifiers modifiers, 
+  unichar *characters
+) {
+  uint32_t modifierKeyState = (modifiers >> 8) & 0xFF;
+  uint32_t deadKeyState = 0;
+  size_t charCount = 0;
+
+  OSStatus status = UCKeyTranslate(
+    keyboardLayout,
+    (uint16_t)virtualKeyCode,
+    kUCKeyActionDown,
+    modifierKeyState,
+    LMGetKbdLast(),
+    kUCKeyTranslateNoDeadKeysBit,
+    &deadKeyState,
+    4 * sizeof(characters[0]),
+    &charCount,
+    characters
+  );
+
+  if (status == noErr && deadKeyState != 0) {
+    status = UCKeyTranslate(
+      keyboardLayout,
+      (uint16_t)virtualKeyCode,
+      kUCKeyActionDown,
+      modifierKeyState,
+      LMGetKbdLast(),
+      kUCKeyTranslateNoDeadKeysBit,
+      &deadKeyState,
+      4 * sizeof(characters[0]),
+      &charCount,
+      characters
+    );
+  }
+
+  if (status == noErr && !iscntrl(characters[0])) {
+    return charCount;
+  } else {
+    return 0;
+  }
+}
+
+CAMLprim value oni2_KeyboardLayoutPopulateCurrentKeymap(value keymap, value Hashtbl_replace) {
```

Comment:
Clever - so the consumer passes a function to collect the keys and mappings?

Reply:
Yep! I was trying to think of the best way to modify an OCaml `Hashtbl.t` in C and that was the idea that I came up with -- the consumer creates a `Hashtbl`, passes it into our `populate` function alongside the `Hashtbl.replace` OCaml callback so that we can modify the table in place from C -- it seemed like a good idea to me, but let me know if you think there's anything wrong with it!

Reply:
I think the idea is great! My only thought is that it seems more general than a `Hashtbl.t` - could potentially use other values, too. I'll add a comment on the signature with more details.